src/Pos/Wallet/Web/Methods/Logic.hs

```diff
@@ -0,0 +1,290 @@
+
+getWAddressBalance :: MonadWalletWebMode m => CWAddressMeta -> m Coin
+getWAddressBalance addr =
+    getBalance <=< decodeCIdOrFail $ cwamId addr
+
+getWAddress
+    :: MonadWalletWebMode m
+    => CachedCAccModifier -> CWAddressMeta -> m CAddress
+getWAddress cachedAccModifier cAddr = do
+    let aId = cwamId cAddr
+    balance <- getWAddressBalance cAddr
+
+    let getFlag customType accessMod = do
+            checkDB <- isCustomAddress customType (cwamId cAddr)
+            let checkMempool = elem aId . map (fst . fst) . toList $
+                               MM.insertions $ accessMod cachedAccModifier
+            return (checkDB || checkMempool)
+    isUsed   <- getFlag UsedAddr camUsed
+    isChange <- getFlag ChangeAddr camChange
+    return $ CAddress aId (mkCCoin balance) isUsed isChange
+
+getAccountAddrsOrThrow
+    :: MonadWalletWebMode m
+    => AddressLookupMode -> AccountId -> m [CWAddressMeta]
+getAccountAddrsOrThrow mode accId =
+    getAccountWAddresses mode accId >>= maybeThrow noWallet
+  where
+    noWallet =
+        RequestError $
+        sformat ("No account with id "%build%" found") accId
+
+getAccount :: MonadWalletWebMode m => CachedCAccModifier -> AccountId -> m CAccount
+getAccount accMod accId = do
+    dbAddrs    <- getAccountAddrsOrThrow Existing accId
+    let modifier   = camAddresses accMod
+    let allAddrIds = gatherAddresses modifier dbAddrs
+    allAddrs <- mapM (getWAddress accMod) allAddrIds
+    balance  <- mkCCoin . unsafeIntegerToCoin . sumCoins <$>
+                mapM getWAddressBalance allAddrIds
+    meta <- getAccountMeta accId >>= maybeThrow noWallet
+    pure $ CAccount (encodeCType accId) meta allAddrs balance
+  where
+    noWallet =
+        RequestError $ sformat ("No account with id "%build%" found") accId
+    gatherAddresses modifier dbAddrs = do
+        let memAddrs = sortedInsertions modifier
+            relatedMemAddrs = filter ((== accId) . addrMetaToAccount) memAddrs
+            -- @|relatedMemAddrs|@ is O(1) while @dbAddrs@ is large
+            unknownMemAddrs = filter (`notElem` dbAddrs) relatedMemAddrs
+        dbAddrs <> unknownMemAddrs
+
+getWallet :: MonadWalletWebMode m => CId Wal -> m CWallet
+getWallet cAddr = do
+    meta       <- getWalletMeta cAddr >>= maybeThrow noWSet
+    wallets    <- getAccounts (Just cAddr)
+    let walletsNum = length wallets
+    balance    <- mkCCoin . unsafeIntegerToCoin . sumCoins <$>
+                     mapM (decodeCCoinOrFail . caAmount) wallets
+    hasPass    <- isNothing . checkPassMatches emptyPassphrase <$> getSKById cAddr
+    passLU     <- getWalletPassLU cAddr >>= maybeThrow noWSet
+    pure $ CWallet cAddr meta walletsNum balance hasPass passLU
+  where
+    noWSet = RequestError $
+        sformat ("No wallet with address "%build%" found") cAddr
+
+-- TODO [CSM-407]: Write smth like `decodeCTypeOrFail` instead of next functions
+decodeCIdOrFail :: MonadThrow m => CId w -> m Address
```

Comment:
Such small functions (without wallet dependencies other than `Pos.Wallet.Web.State` and `Pos.Wallet.Web.ClientTypes`) should belong to `Pos.Wallet.Web.Util`, in my opinion (or even in the aforementioned `Pos.Wallet.Web.ClientTypes.Functions`)